File: sdk/elixir/examples/remote.exs
# RemotePlayer + RemoteController tour over the remote-control WebSocket.
#
#   ROCKSKY_ERL_PATH=../erlang ROCKSKY_TOKEN=<jwt> mix run examples/remote.exs
#
# Needs the native lib built with the remote-player feature (../erlang/build-core.sh)
# and a Rocksky access token (JWT) in $ROCKSKY_TOKEN.

token =
  case System.get_env("ROCKSKY_TOKEN") do
    nil ->
      IO.puts("set ROCKSKY_TOKEN to a Rocksky access token")
      System.halt(1)

    t ->
      t
  end

# --- player: advertise what we're "playing" and react to commands ---
player = Rocksky.RemotePlayer.connect(token, "Elixir Player")

Rocksky.RemotePlayer.listen(player, %{
  play: fn -> IO.puts("player: play") end,
  pause: fn -> IO.puts("player: pause") end,
  next: fn -> IO.puts("player: next") end,
  previous: fn -> IO.puts("player: previous") end,
  seek: fn ms -> IO.puts("player: seek #{ms}ms") end,
  enqueue: fn cmd -> IO.puts("player: enqueue #{length(cmd["tracks"])} track(s)") end
})

{:ok, _} =
  Rocksky.RemotePlayer.set_now_playing(player, %{
    "title" => "Chaser",
    "artist" => "Calibro 35",
    "album" => "Jazzploitation",
    "albumArtist" => "Calibro 35",
    "durationMs" => 182_320,
    "elapsedMs" => 0,
    "isPlaying" => true
  })

{:ok, _} = Rocksky.RemotePlayer.set_status(player, "playing")

{:ok, _} =
  Rocksky.RemotePlayer.set_queue(
    player,
    [%{"title" => "Chaser", "artist" => "Calibro 35", "durationMs" => 182_320}],
    0
  )

# --- controller: observe devices and drive the primary ---
ctl = Rocksky.RemoteController.connect(token, "Elixir Controller")

Rocksky.RemoteController.listen(ctl, %{
  devices: fn e -> IO.puts("controller: #{length(e["devices"])} device(s)") end,
  now_playing: fn e -> IO.puts("controller: now playing on #{e["deviceId"]}") end
})

# broadcast a pause to every device (nil target), then seek the player.
{:ok, _} = Rocksky.RemoteController.pause(ctl)
{:ok, _} = Rocksky.RemoteController.seek(ctl, nil, 42_000)

IO.puts("listening for 10s...")
Process.sleep(10_000)

Rocksky.RemoteController.disconnect(ctl)
Rocksky.RemotePlayer.disconnect(player)
IO.puts("done")

File: sdk/python/examples/remote.py
"""Remote player + controller tour (needs a real Rocksky access token).

    uv run python examples/remote.py <ACCESS_TOKEN>

A RemotePlayer advertises what "this device" is playing and reacts to commands
from a controller; a RemoteController lists the user's players, picks a primary,
observes their state, and sends commands. Both use a background poll thread —
register handlers with ``.on(event, fn)`` and call ``.listen()``.

Needs the native lib — run ./build-core.sh first (or it ships with the wheel).
"""

import sys
import time

from rocksky import (
    RemoteController,
    RemoteNowPlaying,
    RemotePlayer,
    RemoteQueueItem,
    RemoteStatus,
)


def main() -> None:
    if len(sys.argv) < 2:
        print("usage: python examples/remote.py <ACCESS_TOKEN>")
        raise SystemExit(2)
    token = sys.argv[1]

    # ── player: expose this process as a controllable device ────────────────
    player = RemotePlayer(token, "Python Example Player")
    player.on("play", lambda: print("[player] play"))
    player.on("pause", lambda: print("[player] pause"))
    player.on("next", lambda: print("[player] next"))
    player.on("previous", lambda: print("[player] previous"))
    player.on("seek", lambda position_ms: print(f"[player] seek -> {position_ms}ms"))
    player.on("queue_jump", lambda index: print(f"[player] queue jump -> {index}"))
    player.on("queue_remove", lambda index: print(f"[player] queue remove -> {index}"))
    player.on("enqueue", lambda cmd: print(f"[player] enqueue {len(cmd.tracks)} track(s), mode={cmd.mode}"))
    player.listen()

    # Advertise a track + queue so controllers have something to show.
    player.set_now_playing(
        RemoteNowPlaying(
            title="Chaser",
            artist="Calibro 35",
            album="Jazzploitation",
            album_artist="Calibro 35",
            duration_ms=182320,
            elapsed_ms=0,
            is_playing=True,
        )
    )
    player.set_status(RemoteStatus.PLAYING)
    player.set_queue(
        [RemoteQueueItem(title="Chaser", artist="Calibro 35", album="Jazzploitation")],
        0,
    )

    # ── controller: observe + drive the user's players ──────────────────────
    controller = RemoteController(token, "Python Example Controller")
    controller.on("devices", lambda e: print(f"[controller] {len(e.devices)} device(s), primary={e.primary_device}"))
    controller.on("now_playing", lambda e: print(f"[controller] {e.device_name}: {e.track.artist} — {e.track.title}"))
    controller.on("status", lambda e: print(f"[controller] {e.device_name}: {e.status}"))
    controller.on("queue", lambda e: print(f"[controller] {e.device_name}: queue index {e.index}"))
    controller.listen()

    # Drive whatever devices exist (None target = broadcast to all).
    time.sleep(1)
    controller.pause(None)
    controller.seek(None, 42000)

    print("listening for 30s — control this player from the Rocksky miniplayer…")
    try:
        time.sleep(30)
    except KeyboardInterrupt:
        pass
    finally:
        controller.disconnect()
        player.disconnect()


if __name__ == "__main__":
    main()
